caps/types.go

```diff
@@ -55,11 +58,26 @@ func (t *Type) Validate(c *Capability) error {
 	if t != c.Type {
 		return fmt.Errorf("capability is not of type %q", t)
 	}
-	// While we don't have any support for type-specific attribute schema,
-	// let's ensure that attributes are totally empty. This will make tests
-	// show that this code is actually being used.
-	if c.Attrs != nil && len(c.Attrs) != 0 {
-		return fmt.Errorf("attributes must be empty for now")
+	// Check that all supported attributes are present
+	for _, attr := range t.RequiredAttrs {
+		if _, ok := c.Attrs[attr]; !ok {
+			return fmt.Errorf("capability lacks required attribute %q", attr)
+		}
+	}
+	// Look for any unexpected attributes
+	if len(t.RequiredAttrs) != len(c.Attrs) {
+		for attr := range c.Attrs {
+			supported := false
```

Comment:
(nitpick, probably over the top this time). Its a bit unfortunate that there is no "Contains()" outside of the testutil package :) I think this would be easier to read by:
```
for attr := range c.Attrs {
			if !Contains(t.RequiredAttrs, attr)  {
				return fmt.Errorf("capability contains unexpected attribute %q", attr)
			}
```
But not for this branch :)